source/common/common/perf_annotation.h

```diff
@@ -0,0 +1,184 @@
+#pragma once
+
+#ifdef ENVOY_PERF_ANNOTATION
+
+#include <chrono>
+#include <cstdint>
+#include <mutex>
+#include <unordered_map>
+
+#include "common/common/utility.h"
+
+#include "absl/strings/string_view.h"
+
+// Performance Annotation system, enabled with
+//   bazel --define=perf_annotation=enabled ...
+// or, in individual .cc files:
+//   #define ENVOY_PERF_ANNOTATION
+// In the absense of such directives, the support classes are built and tested.
+// However, the macros for instrumenting code for performance analysis will expand
+// to nothing.
+//
+// See also: https://github.com/LLNL/Caliper -- it may be worth integrating with
```

Comment:
This methodology is also similar to KStats used at VMware for the VMM, see https://labs.vmware.com/vmtj/methodology-for-performance-analysis-of-vmware-vsphere-under-tier-1-applications. We wrote a bit about how we used this in https://dl.acm.org/citation.cfm?id=1899945&dl=ACM&coll=DL as well.

Reply:
added more comments for further reading.